x/auth/vesting/types/vesting_account.go

```diff
@@ -611,3 +615,297 @@ func marshalYaml(i interface{}) (interface{}, error) {
+		grantVestingPeriods: grantVestingPeriods,
+		grantCoins:          grantCoins,
+	}
+}
+
+func (cga clawbackGrantAction) AddToAccount(ctx sdk.Context, rawAccount exported.VestingAccount) error {
+	cva, ok := rawAccount.(*ClawbackVestingAccount)
+	if !ok {
+		return fmt.Errorf("expected *ClawbackVestingAccount, got %T", rawAccount)
+	}
+	if cga.funderAddress != cva.FunderAddress {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s can only accept grants from account %s",
+			rawAccount.GetAddress(), cva.FunderAddress)
+	}
+	cva.addGrant(ctx, cga.grantStartTime, cga.grantLockupPeriods, cga.grantVestingPeriods, cga.grantCoins)
+	return nil
+}
+
+func (va *ClawbackVestingAccount) AddGrant(ctx sdk.Context, action exported.AddGrantAction) error {
+	return action.AddToAccount(ctx, va)
+}
+
+func (va *ClawbackVestingAccount) addGrant(ctx sdk.Context, grantStartTime int64, grantLockupPeriods, grantVestingPeriods []Period, grantCoins sdk.Coins) {
+	// modify schedules for the new grant
+	newLockupStart, newLockupEnd, newLockupPeriods := DisjunctPeriods(va.GetStartTime(), grantStartTime, va.LockupPeriods, grantLockupPeriods)
+	newVestingStart, newVestingEnd, newVestingPeriods := DisjunctPeriods(va.GetStartTime(), grantStartTime,
+		va.GetVestingPeriods(), grantVestingPeriods)
+	if newLockupStart != newVestingStart {
+		panic("bad start time calculation")
+	}
+	va.StartTime = newLockupStart
+	va.EndTime = max64(newLockupEnd, newVestingEnd)
+	va.LockupPeriods = newLockupPeriods
+	va.VestingPeriods = newVestingPeriods
+	va.OriginalVesting = va.OriginalVesting.Add(grantCoins...)
+}
+
+// GetUnlockedOnly returns the unlocking schedule at blockTIme.
+func (va ClawbackVestingAccount) GetUnlockedOnly(blockTime time.Time) sdk.Coins {
+	return ReadSchedule(va.GetStartTime(), va.EndTime, va.LockupPeriods, va.OriginalVesting, blockTime.Unix())
+}
+
+// GetVestedOnly returns the vesting schedule at blockTime.
+func (va ClawbackVestingAccount) GetVestedOnly(blockTime time.Time) sdk.Coins {
+	return ReadSchedule(va.GetStartTime(), va.EndTime, va.VestingPeriods, va.OriginalVesting, blockTime.Unix())
+}
+
+// computeClawback removes all future vesting events from the account,
+// returns the total sum of these events. When removing the future vesting events,
+// the lockup schedule will also have to be capped to keep the total sums the same.
+// (But future unlocking events might be preserved if they unlock currently vested coins.)
+// If the amount returned is zero, then the returned account should be unchanged.
+// Does not adjust DelegatedVesting
+// TODO: Rename this function, its doing more than computing a clawback, its altering the struct.
+func (va *ClawbackVestingAccount) computeClawback(clawbackTime int64) sdk.Coins {
+	// Compute the truncated vesting schedule and amounts.
+	// Work with the schedule as the primary data and recompute derived fields, e.g. OriginalVesting.
+	vestTime := va.GetStartTime()
+	totalVested := sdk.NewCoins()
```

Comment:
Do we need to compute both `totalVested` and `totalUnvested`. Is there a `total` of both stored in the base account? If yes, we can only compute `totalVested`, break out of the loop early and just subtract `totalVested` from `total` to get `totalUnvested`